src/frontend/apps/impress/src/features/docs/doc-management/libs/CollaborationProvider.ts

```diff
@@ -0,0 +1,333 @@
+      }, 5000);
+    } else if (status === WebSocketStatus.Connected) {
+      this.setPollDefaultValues();
+    }
+
+    super.onStatus({ status });
+  }
+
+  public initPolling() {
+    if (this.isLongPollingStarted || !this.isWebsocketFailed) {
+      return;
+    }
+
+    this.isLongPollingStarted = true;
+    void this.pollSync(true);
+    this.initCollaborationSSE();
+  }
+
+  protected toPollUrl(endpoint: string): string {
+    let pollUrl = this.url.replace('ws:', 'http:');
+    if (pollUrl.includes('wss:')) {
+      pollUrl = pollUrl.replace('wss:', 'https:');
+    }
+
+    pollUrl = pollUrl.replace('/ws/', '/ws/poll/' + endpoint + '/');
+
+    // To have our requests not cached
+    return `${pollUrl}&${Date.now()}`;
+  }
+
+  protected isDuplicateAwareness(message64: string): boolean {
+    if (this.treatedAwarenessMessage === message64) {
+      return true;
+    }
+
+    this.treatedAwarenessMessage = message64;
+
+    return false;
+  }
+
+  /**
+   * Outgoing message event
+   *
+   * Sent to the server the message to
+   * be sent to the other users
+   */
+  public async onPollOutgoingMessage({ message }: onOutgoingMessageParameters) {
+    if (!this.isWebsocketFailed || !this.canEdit) {
+      return;
+    }
+
+    const message64 = Buffer.from(message.toUint8Array()).toString('base64');
+    if (this.isDuplicateAwareness(message64)) {
+      return;
+    }
+
+    try {
+      const { updated } = await pollOutgoingMessageRequest({
+        pollUrl: this.toPollUrl('message'),
+        message64,
+      });
+
+      if (!updated) {
+        await this.pollSync();
+      }
+    } catch (error: unknown) {
+      if (isAPIError(error)) {
+        // The user is not allowed to send messages
+        if (error.status === 403) {
+          this.off('outgoingMessage', this.onPollOutgoingMessage.bind(this));
+          this.canEdit = false;
+        }
+      }
+    }
+  }
+
+  /**
+   * EventSource is a API for opening an HTTP
+   * connection for receiving push notifications
+   * from a server in real-time.
+   * We use it to sync the document with the server
+   */
+  protected initCollaborationSSE() {
+    if (!this.isWebsocketFailed) {
+      return;
+    }
+
+    this.sse = new EventSource(this.toPollUrl('message'), {
+      withCredentials: true,
+    });
+
+    this.sse.onmessage = (event) => {
+      const { updatedDoc64, stateFingerprint, awareness64 } = JSON.parse(
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+        event.data,
+      ) as {
+        updatedDoc64?: string;
+        stateFingerprint?: string;
+        awareness64?: string;
+      };
+
+      if (awareness64) {
+        if (this.isDuplicateAwareness(awareness64)) {
+          return;
+        }
+
+        this.treatedAwarenessMessage = awareness64;
+        const awareness = Buffer.from(awareness64, 'base64');
+
+        this.onMessage({
+          data: awareness,
+        } as MessageEvent);
+      }
+
+      if (updatedDoc64) {
```

Comment:
Why do we have our own logic to apply the message? Isn't it enough (and easier), to let the `onMessage` function handle this for all cases? Or is there a reason that doesn't work?

---

Zoomed out; I'm a little concerned by all the manual Yjs operations you have to do. I was hoping you could just use the existing sync-protocol (and code that handles that), but only over a different transport layer. My guess is you ran into issues with this and came up with some workarounds? That does make the code a little more difficult to review (especially when I don't have the context of why which workarounds are necessary).